### ISSUE TITLE
[breaking change]: Deprecation and discontinuation of standalone Dart tools

### ISSUE THREAD

**Issue by mit-mit** [dart-lang/sdk#46100]
## Background

Historically the Dart team has offered a number of smaller developer tools for things like formatting code (`dartfmt`), analyzing code (`dartanalyzer`), etc. In Dart 2.10 (October 2020) we introduced a new unified Dart developer tool, the [`dart` tool](https://dart.dev/tools/dart-tool).

## Intended change

As we discussed at launch in the [blog post](https://medium.com/dartlang/announcing-dart-2-10-350823952bd5), we plan to first deprecate and then discontinue the smaller tools in favor of the new combined tool. The this meta-bug presents an overview of the status of this migration. As this is considered breaking, it will happen in a staged fashion over time, when each individual tool first becoming deprecated in one stable release (here it will still exist but print a warning), and then fully discontinued (here it will no longer be in the SDK).

| Historical tool | `dart` replacement                                       | Deprecation | Discontinuation |
|-----------------|----------------------------------------------------------|-------------|-----------------|
| `stagehand`     |  `dart create`                                           | [2.14](https://github.com/dart-lang/stagehand/issues/671)        | ✅ 2.14*  |
| `dartfmt`       | [`dart format`](https://dart.dev/tools/dart-format)      | [2.14](https://github.com/dart-lang/dart_style/issues/986)        | ✅ [2.15](https://github.com/dart-lang/dart_style/issues/986)            |
| `dart2native`   | [`dart compile exe`](https://dart.dev/tools/dart-compile)    | [2.14](https://github.com/dart-lang/sdk/commit/cac00e9d956a6f7ef28628989912d971f6b908d4)        | ✅ [2.15](https://github.com/dart-lang/sdk/commit/6c5fb84716b1f257b170351efe8096fe2af2809b)            |
| `dart2js`       | [`dart compile js`](https://dart.dev/tools/dart-compile)    | [2.17](https://github.com/dart-lang/sdk/commit/8415b70e75b1d5bbe8251fa6a9eab2d970cf9eec)         | ✅ [2.18](https://github.com/dart-lang/sdk/commit/69249df50bcc7a0489176efd3fd79fff018f1b91)             |
| `dartdevc`      | none                              | [2.17](https://github.com/dart-lang/sdk/commit/5173fd2d224f669fd8d0a1d21adbfd6187d10f53)         | ✅ [2.18](https://github.com/dart-lang/sdk/commit/69249df50bcc7a0489176efd3fd79fff018f1b91)             |
| `dartanalyzer` & [`package:analyzer_cli`](https://github.com/dart-lang/sdk/tree/master/pkg/analyzer_cli)  | [`dart analyze`](https://dart.dev/tools/dart-analyze)    | [2.16](https://github.com/dart-lang/sdk/commit/f7af5c5256ee6f3a167f380722b96e8af4360b46)         | ✅ [2.18](https://github.com/dart-lang/sdk/issues/48457)             |
| `dartdoc`       | `dart doc`                                                      | [2.16](https://github.com/dart-lang/sdk/issues/44610)         | ✅ [2.17](https://dart-review.googlesource.com/c/sdk/+/228647)             |
| `pub`           | [`dart pub`](https://dart.dev/tools/dart-pub)            | [2.15](https://github.com/dart-lang/pub/issues/2736)         | ✅ [2.17](https://dart-review.googlesource.com/c/sdk/+/234283)             |

*Notes*: 

- We currently have no plans to deprecate [`dartaotruntime`](https://dart.dev/tools/dartaotruntime).
- `stagehand` will go directly to discontinuation as it's a `pub global` tool on pub.dev, and not a SDK distributed tool like the rest. This means it can still be invoked (with `dart pub global run stagehand`) even after it's been discontinued.
- `dartdevc` will be discontinued as a publicly visible tool. The tool was never intended or supported as a stable command line tool for end users. The underlying binary for this compiler will still be shipped with the SDK to be used by build systems tools (such as bazel, `webdev`, and `flutter_tools`). However, it will no longer be exposed to be used by developers directly on the command line.

## Rationale

With the new unified tool the existing tools are no longer needed. To reduce maintenance, remove confusion for new users, and to ensure we offer a single high-quality toolset the old commands are being removed.

Should you have any issues with missing functionality in the new `dart`-based tools, please file an issue.

## Migration

To migrate, switch to using the [`dart` tool](https://dart.dev/tools/dart-tool), and the subcommands listed in the above table.

**Comment by jonahwilliams** [comment]
I can begin switching the flutter usages of the packages with specified deprecation dates. For some of these, it seems like there is still a significant gap in functionality (analysis server usage with dart analyze, or the various compile options for dart2js). Is there a plan to resolve this before announcing deprecation dates?

**Comment by sigmundch** [comment]
> ... significant gap in functionality (... various compile options for dart2js)

I can at least speak for the dart2js side, yes. We will ensure there isn't any gap in our case. It's very likely that we'll even preserve all existing flags, but we'll confirm once the plan is finalized.

**Comment by mit-mit** [comment]
The items marked as deprecated in 2.14 in [the table](https://github.com/dart-lang/sdk/issues/46100#issue-897970252) are considered "feature complete" and we're not aware of any things missing from them.

For `dart analyze` it's fairly complete, and we'd like to hear more input on what might be missing. We've migrated a bunch of uses of `dartanalyzer` in various Dart repos already.

**Comment by kleinpetr** [comment]
what is the replacement for the **web-angular** which is in now in stagehand?

**Comment by aadarshadhakalg** [comment]
As angular-dart is also being discontinued, most probably dart create will not include web-angular template. But Is the google team planning to release a separate CLI tool for angular-dart or there are other existing alternatives?

**Comment by mit-mit** [comment]
`stagehand` will continue to work as today; it just won't see any future work -- so that can still be used to create web-angular projects.

**Comment by jonahwilliams** [comment]
@mit-mit I'm thinking of functionality like the lsp mode that the analysis server supports, unless that snapshot won't be removed.

**Comment by bkonyi** [comment]
The analysis server snapshot won't be going anywhere any time soon since we don't have a replacement for it in the CLI.

**Comment by jonahwilliams** [comment]
ack, I actually forgot about dartanalyzer, but that one makes more sense 😅

**Comment by munificent** [comment]
As of [this CL](https://dart-review.googlesource.com/c/sdk/+/202144), dartfmt is now printing a deprecation notice.

**Comment by simolus3** [comment]
In the 2.16 release notice, it was mentioned that the standalone `dartdoc` command is going away in 2.17.

I'm currently using the options `--rel-canonical-prefix` and `--link-to-source-revision` when invoking `dartdoc`. `dart doc` doesn't appear to accept those options, what's the replacement for them?

**Comment by leafpetersen** [comment]
As discussed in various meetings, support for analyzing the Dart SDK core libraries doesn't seem to be available with the new command form.  From the command line, currently:
```
xcodebuild/ReleaseX64/dart-sdk/bin/dartanalyzer  --dart-sdk=/Users/leafp/src/dart-repo/sdk/sdk --sdk-warnings  /Users/leafp/src/dart-repo/sdk/sdk/lib/
```
works but the corresponding `dart analyze` command does not.

**Comment by simolus3** [comment]
>  I'm currently using the options `--rel-canonical-prefix` and `--link-to-source-revision` when invoking `dartdoc`. `dart doc` doesn't appear to accept those options, what's the replacement for them?

To anyone else also stumbling across this, my solution was to add a `dev_dependency` on `dartdoc` and then use `dart run dartdoc` instead of the more limited `dart doc` command.

**Comment by devoncarew** [comment]
@simolus3 - can you open an issue against https://github.com/dart-lang/dartdoc for the dartdoc CLI flags you're missing? And, have you tried configuring dartdoc via the dartdoc_options.yaml file (https://github.com/dart-lang/dartdoc#dartdoc_optionsyaml)? That allows you to tweak more parameters than are available through `dart doc`.

**Comment by sigmundch** [comment]
@jakemac53 - yes, the goal is still to support their use from build_web_compilers. For now we intend to keep the snapshots, but there have been discussions about combining the snapshots into the CLI snapshot sometime in the future (but that's not a short term goal).

To avoid the warning, using `dart compile js` will work (that will also be safe in the future if dart2js gets merged into the Dart CLI). Alternatively, we can also pass a special flag to the dart2js snapshot to indicate that you are one of the use cases we support (that was our intended design at first).

As for DDC, the snapshot will remain available for foreseeable future. Note that DDC should not give you a warning if you are using the snapshot (only the shell script does today).

**Comment by jakemac53** [comment]
Thanks, I will go ahead and land the change to  use `dart compile js` and leave the rest as is for now. So you should be unblocked soon to merge the snapshot with the CLI snapshot when you are able, that should probably help sdk size a fair bit :).
